packages/aws-cdk/lib/api/deploy-stack.ts

```diff
@@ -177,42 +188,57 @@ async function getStackOutputs(cfn: aws.CloudFormation, stackName: string): Prom
 }
 
 /**
- * Prepares the body parameter for +CreateChangeSet+, putting the generated CloudFormation template in the toolkit-provided
- * S3 bucket if present, otherwise using in-line template argument. If no +ToolkitInfo+ is provided and the template is
- * larger than 50,200 bytes, an +Error+ will be raised.
+ * Prepares the body parameter for +CreateChangeSet+.
+ *
+ * If the template is small enough to be inlined into the API call, just return
+ * it immediately.
+ *
+ * Otherwise, add it to the asset manifest to get uploaded to the staging
+ * bucket and return its coordinates. If there is no staging bucket, an error
+ * is thrown.
  *
  * @param stack     the synthesized stack that provides the CloudFormation template
- * @param sdk     an AWS SDK to use when interacting with S3
  * @param toolkitInfo information about the toolkit stack
  */
-async function makeBodyParameter(stack: cxapi.CloudFormationStackArtifact, toolkitInfo?: ToolkitInfo): Promise<TemplateBodyParameter> {
+async function makeBodyParameter(
+  stack: cxapi.CloudFormationStackArtifact,
+  assetManifest: AssetManifestBuilder,
+  toolkitInfo?: ToolkitInfo): Promise<TemplateBodyParameter> {
   const templateJson = toYAML(stack.template);
-  if (toolkitInfo) {
-    const s3KeyPrefix = `cdk/${stack.id}/`;
-    const s3KeySuffix = '.yml';
-    const { key } = await toolkitInfo.uploadIfChanged(templateJson, {
-      s3KeyPrefix, s3KeySuffix, contentType: 'application/x-yaml'
-    });
-    const templateURL = `${toolkitInfo.bucketUrl}/${key}`;
-    debug('Stored template in S3 at:', templateURL);
-    return { TemplateURL: templateURL };
-  } else if (templateJson.length > LARGE_TEMPLATE_SIZE_KB * 1024) {
+
+  if (templateJson.length <= LARGE_TEMPLATE_SIZE_KB * 1024) {
+    return { TemplateBody: templateJson };
+  }
+
+  if (!toolkitInfo) {
     error(
       `The template for stack "${stack.displayName}" is ${Math.round(templateJson.length / 1024)}KiB. ` +
       `Templates larger than ${LARGE_TEMPLATE_SIZE_KB}KiB must be uploaded to S3.\n` +
       'Run the following command in order to setup an S3 bucket in this environment, and then re-deploy:\n\n',
       colors.blue(`\t$ cdk bootstrap ${stack.environment!.name}\n`));
 
     throw new Error(`Template too large to deploy ("cdk bootstrap" is required)`);
-  } else {
-    return { TemplateBody: templateJson };
   }
+
+  const templateHash = contentHash(templateJson);
+  const key = `cdk/${stack.id}/${templateHash}.yml`;
```

Comment:
I was expecting this to happen in the framework and not in the CLI, so that templates are just treated like any other asset.

Reply:
Well yeah, but similar to the last query: this is not true yet, I will add an additional `if` in the PR that introduces asset manifests from the framework side.